examples/pytorch/sage.py

```diff
@@ -0,0 +1,229 @@
+        )  # Number of output nodes of layer 2
+        return context
+
+    def _loss_inner(
+        self, context: QueryOutput
+    ) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+        """Cross entropy loss for a list of nodes."""
+        if isinstance(context, dict):
+            labels = context["label"].squeeze()
+        elif isinstance(context, torch.Tensor):
+            labels = context.squeeze()  # type: ignore
+        else:
+            raise TypeError("Invalid input type.")
+        scores = self.get_score(context)
+        labels = labels.to(torch.int64)
+        return (
+            self.xent(
+                scores,
+                labels,
+            ),
+            scores.argmax(dim=1),
+            labels,
+        )
+
+    def forward(
+        self, context: QueryOutput
+    ) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+        """Return cross entropy loss."""
+        return self._loss_inner(context)
+
+    # type: ignore[override]
+    def get_score(self, context: dict) -> torch.Tensor:
+        """Generate scores for a list of nodes."""
+        embeds = self.get_embedding(context)
+        scores = torch.matmul(embeds, self.weight)
+        return scores
+
+    def metric_name(self):
+        """Metric used for training."""
+        return self.metric.name()
+
+    # type: ignore[override]
+    def get_embedding(self, context: dict) -> torch.Tensor:
+        """Generate embedding."""
+        out_1 = context["out_1"][0]
+        out_2 = context["out_2"][0]
+        try:
+            out_1 = out_1[0]
+            out_2 = out_2[0]
+        except IndexError:
+            pass
+        edges_1 = self.build_edges_tensor(out_1, self.fanouts[0])  # Edges for 1st layer
+        x1 = self.convs[0](
+            context["x0"].reshape((-1, context["x0"].shape[-1])), edges_1
+        )[
+            :out_1, :
+        ]  # Output of 1st layer (cut out 2 hop nodes)
+        x1 = F.relu(x1)
+        edges_2 = self.build_edges_tensor(out_2, self.fanouts[1])  # Edges for 2nd layer
+        x2 = self.convs[1](x1, edges_2)[
+            :out_2, :
+        ]  # Output of second layer (nodes for which loss is computed)
+        x2 = F.relu(x2)
+        return x2
+
+
+def train_func(config: dict):
+    """Training loop for ray trainer."""
+
+    model = PTGSupervisedGraphSage(
+        num_classes=config["num_classes"],
+        label_idx=config["label_idx"],
+        label_dim=config["label_dim"],
+        feature_dim=config["feature_dim"],
+        feature_idx=config["feature_idx"],
+        edge_type=0,
+        fanouts=[10, 10],
+    )
+    model.train()
+
+    optimizer = torch.optim.Adam(model.parameters(), lr=0.01)
+    g = MemoryGraph(config["data_dir"], partitions=[0])
+    train_dataset = np.loadtxt(f"{config['data_dir']}/train.nodes", dtype=np.int64)
+    eval_batch = np.loadtxt(f"{config['data_dir']}/test.nodes", dtype=np.int64)
+
+    for epoch in range(config["num_epochs"]):
```

Comment:
Why don't we hardcode these arguments instead of putting them in train_func config?

Reply:
Names are more descriptive. Also easier to refactor if needed, e.g. some cusomt graphs might have 500dim features.